### PR TITLE
Fix issue in beta reduction and add support in ALF

### PR DESCRIPTION
**Pull request by ajreynol**
This adds support for ProofRewriteRule::BETA_REDUCE in ALF.

When formalizing this rule, three issues arose:

(1) We do not properly model variables from define-fun, thus `beta-reduce` was not properly checkable in the ALF signature.  This PR changes our smt2 parser to use canonical variables, similar to how the parser for quantified formulas was updated here: https://github.com/cvc5/cvc5/pull/10556.

(2) When fixing this discrepancy, a bug was uncovered in the UF rewriter. In particular, the cvc5 rewriter does *not* rewrite terms that appear in operators (this occurs rarely, only when there are beta-redexes or we have an HO logic).  This meant that it was theoretically possible for a lambda with variable shadowing to appear in an APPLY_UF term, thus leading to a bogus beta reduction when rewritten.  This issue could only appear in raw API usages where a user could manually construct a lambda with shadowing and put it in an APPLY_UF.

This PR fixes this issue by ensuring lambdas are rewritten prior to performing beta reduction. At the moment, this requires a backwards reference to the rewriter from the UF rewriter (this is not an ideal practice, but a similar pattern is used in some of our other theory rewriters).

(3) Our shadow elimination was not considering the case where a variable occurs multiple times in the same quantifier prefix. This is also fixed in this PR.

A more permanent solution will be to rewrite inside of operators by default in the core rewriter, although this is a major change.

This change should be performance tested as it impacts our default handling of `define-fun`.

### CODE REVIEW

**Comment by ajreynol** [comment]
This had neglible performance impact on quantifier-free strings+arithmetic+UF. Will report more logics as I test more, for now, I am confident this PR should go in.

File: src/expr/elim_shadow_converter.cpp [hunk 1]
```diff
@@ -91,7 +94,31 @@ Node ElimShadowNodeConverter::eliminateShadow(const Node& q)
   ElimShadowNodeConverter esnc(nm, q);
   // eliminate shadowing in all children
   std::vector<Node> children;
-  children.push_back(q[0]);
+  // drop duplicate variables
```
Comment (HanielB) [inline src/expr/elim_shadow_converter.cpp:97]: If you are changing the quantifier prefix, please add that to the documentation.

File: src/theory/uf/theory_uf_rewriter.cpp [hunk 2]
```diff
@@ -175,6 +200,10 @@ Node TheoryUfRewriter::rewriteViaRule(ProofRewriteRule id, const Node& n)
       {
         return Node::null();
       }
+      // Note that we do not check for variable shadowing in the lambda here.
+      // This rule will only be used to capture valid instances of beta
```
Comment (HanielB) [inline src/theory/uf/theory_uf_rewriter.cpp:204]: "capture" and "beta reduction" in the same sentence immediately lead one to think of variable capture. Please use a different word. :)

**Review by HanielB — APPROVED** [review]
LGTM, just minor comments

**Pull request merged**